bundles/org.openhab.core.io.transport.mqtt/src/main/java/org/eclipse/smarthome/io/transport/mqtt/internal/Subscription.java

```diff
@@ -0,0 +1,91 @@
+/**
+ * Copyright (c) 2010-2019 Contributors to the openHAB project
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.eclipse.smarthome.io.transport.mqtt.internal;
+
+import java.util.Collection;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.smarthome.io.transport.mqtt.MqttMessageSubscriber;
+
+import com.hivemq.client.mqtt.mqtt3.message.publish.Mqtt3Publish;
+import com.hivemq.client.mqtt.mqtt5.message.publish.Mqtt5Publish;
+
+/**
+ * This class keeps track of all the subscribers to a specific topic.
+ * <p>
+ * <b>Retained</b> messages for the topic are stored so they can be replayed to new subscribers.
+ *
+ * @author Jochen Klein - Initial contribution
+ */
+@NonNullByDefault
+public class Subscription {
+
+    private Map<String, byte[]> retainedMessages = new ConcurrentHashMap<>();
```

Comment:
If we have one `Subscription` per topic it should be enough to store the message, the topic should be the same for all messages.

Reply:
No, the topic can be a wildcard topic `homeassistant/#` so there can by multiple messages.

Reply:
Good point